### PR TITLE
chore: re-write connect to use driver

### LINKED ISSUE #1177: [internal]: Rewrite BrowserType.connect to use server side connect
fixes #1071

### PR DESCRIPTION
**Pull request by rwoll**
Remove dependency on websocket lib, and use shared connect impl.

This aligns the implementation with .NET and Java by moving connect over to the server side.

Resolves #1177.

### CODE REVIEW

File: playwright/_impl/_connection.py [hunk 1]
```diff
@@ -205,38 +204,18 @@ def local_utils(self) -> "LocalUtils":
...
-        self.cleanup()
-
-    def cleanup(self) -> None:
-        if self._init_task and not self._init_task.done():
-            self._init_task.cancel()
-        for ws_connection in self._child_ws_connections:
-            ws_connection._transport.dispose()
-        self.emit("close")
+        asyncio.create_task(init())
+        # self.emit("close")
```
Comment (rwoll) [inline playwright/_impl/_connection.py:218]:
```suggestion
```

File: playwright/sync_api/_context_manager.py [hunk 2]
```diff
@@ -94,7 +101,9 @@ def start(self) -> SyncPlaywright:
         return self.__enter__()
 
     def __exit__(self, *args: Any) -> None:
-        self._connection.stop_sync()
+        # self._connection.stop_sync()
```
Comment (rwoll) [inline playwright/sync_api/_context_manager.py:104]:
```suggestion
```

File: playwright/sync_api/_context_manager.py [hunk 3]
```diff
@@ -64,16 +66,21 @@ def __enter__(self) -> SyncPlaywright:
         # until the end of times. We will pass control to that fiber every time we
         # block while waiting for a response.
         def greenlet_main() -> None:
-            self._loop.run_until_complete(self._connection.run_as_sync())
+            self._loop.run_until_complete(self._transport.run())
 
-        dispatcher_fiber = greenlet(greenlet_main)
+        # dispatcher_fiber = greenlet(greenlet_main)
```
Comment (rwoll) [inline playwright/sync_api/_context_manager.py:71]:
```suggestion
```

File: playwright/_impl/_transport.py [hunk 4]
```diff
@@ -52,91 +45,47 @@ def _get_stderr_fileno() -> Optional[int]:
...
     ) -> None:
-        super().__init__(loop)
         self._stopped = False
         self._driver_executable = driver_executable
+        self._loop = loop
+
+        self.on_message: Callable[[ParsedMessagePayload], None]
+        self._stopped_event = asyncio.Event()
+        # self.on_message: Callable[[ParsedMessagePayload], None] = lambda _: None
+        # self.on_error_future: asyncio.Future = loop.create_future()
```
Comment (rwoll) [inline playwright/_impl/_transport.py:59]:
```suggestion
```

File: playwright/_impl/_browser_type.py [hunk 5]
```diff
@@ -181,59 +179,50 @@ async def connect_over_cdp(
...
 
-        transport = WebSocketTransport(
-            self._connection._loop, ws_endpoint, headers, slow_mo
-        )
         connection = Connection(
             self._connection._dispatcher_fiber,
             self._connection._object_factory,
-            transport,
-            self._connection._loop,
+            asyncio.get_running_loop(),
```
Comment (mxschmitt) [inline playwright/_impl/_browser_type.py:195]:
```suggestion
            self._connection._loop,
```

why not that?

File: playwright/_impl/_browser_type.py [hunk 6]
```diff
@@ -181,59 +179,50 @@ async def connect_over_cdp(
...
+            except Exception:
+                if not browser.is_connected:
+                    pipe_closed()
+                raise
+
+        connection.on_message = on_message
+        pipe.on("message", connection.dispatch)
+        await connection.init()
+        playwright = await connection.playwright_future
+        browser: Browser = from_channel(playwright._initializer["preLaunchedBrowser"])
```
Comment (mxschmitt) [inline playwright/_impl/_browser_type.py:223]:
```suggestion
        playwright._set_selectors(self._playwright.selectors)
        browser: Browser = from_channel(playwright._initializer["preLaunchedBrowser"])
```

File: playwright/_impl/_browser_type.py [hunk 7]
```diff
@@ -181,59 +179,50 @@ async def connect_over_cdp(
...
+        browser._is_remote = True
```
Comment (mxschmitt) [inline playwright/_impl/_browser_type.py:224]:
```suggestion
        browser._set_browser_type(self)
        browser._is_remote = True
```

**Review by mxschmitt — APPROVED** [review]

File: playwright/async_api/_context_manager.py [hunk 8]
```diff
@@ -25,30 +25,29 @@
...
-            {self._connection._transport.on_error_future, playwright_future},
-            return_when=asyncio.FIRST_COMPLETED,
-        )
-        if not playwright_future.done():
-            playwright_future.cancel()
-        playwright = AsyncPlaywright(next(iter(done)).result())
+
+        self._connection = Connection(None, create_remote_object, loop)
+        self._transport = PipeTransport(loop, compute_driver_executable())
+        self._connection = connection = Connection(None, create_remote_object, loop)
```
Comment (rwoll) [inline playwright/async_api/_context_manager.py:35]:
```suggestion
        self._transport = PipeTransport(loop, compute_driver_executable())
        self._connection = connection = Connection(None, create_remote_object, loop)
```

File: playwright/async_api/_context_manager.py [hunk 9]
```diff
@@ -25,30 +25,29 @@
...
+        self._transport = transport = PipeTransport(loop, compute_driver_executable())
+        connection.on_message = transport.send
+        transport.on_message = connection.dispatch
+
+        await transport.connect()
```
Comment (dgozman) [inline playwright/async_api/_context_manager.py:40]: Some async method below should throw if transport misbehaves. Let's double check.

File: playwright/sync_api/_context_manager.py [hunk 10]
```diff
@@ -64,16 +66,19 @@ def __enter__(self) -> SyncPlaywright:
...
-            PipeTransport(self._loop, compute_driver_executable()),
-            self._loop,
+        self._dispatcher_fiber = dispatcher_fiber = greenlet(greenlet_main)
+        self._connection = connection = Connection(
+            dispatcher_fiber, create_remote_object, self._loop
+        )
+        self._transport = transport = PipeTransport(
+            self._loop, compute_driver_executable()
         )
+        connection.on_message = transport.send
```
Comment (dgozman) [inline playwright/sync_api/_context_manager.py:78]: Perhaps just use `self._transport` and alike?

File: playwright/_impl/_browser_type.py [hunk 11]
```diff
@@ -181,59 +179,50 @@ async def connect_over_cdp(
...
+        slowMo: float = None,
         headers: Dict[str, str] = None,
     ) -> Browser:
-        if timeout is None:
-            timeout = 30000
-
+        pipe: JsonPipe = from_channel(
+            await self._channel.send("connect", locals_to_params(locals()))
+        )
         headers = {**(headers if headers else {}), "x-playwright-browser": self.name}
```
Comment (dgozman) [inline playwright/_impl/_browser_type.py:190]: This should be above `channel.send`.

File: playwright/async_api/_context_manager.py [hunk 12]
```diff
@@ -25,30 +25,29 @@
...
-
-        done, _ = await asyncio.wait(
-            {self._connection._transport.on_error_future, playwright_future},
-            return_when=asyncio.FIRST_COMPLETED,
-        )
-        if not playwright_future.done():
-            playwright_future.cancel()
-        playwright = AsyncPlaywright(next(iter(done)).result())
+
+        self._connection = Connection(None, create_remote_object, loop)
```
Comment (dgozman) [inline playwright/async_api/_context_manager.py:33]: Should we pass on_close here?

File: playwright/_impl/_browser_type.py [hunk 13]
```diff
@@ -181,59 +179,50 @@ async def connect_over_cdp(
...
-        browser._should_close_connection_on_close = True
-
-        def handle_transport_close() -> None:
+        def pipe_closed() -> None:
             for context in browser.contexts:
                 for page in context.pages:
                     page._on_close()
                 context._on_close()
             browser._on_close()
-            connection.cleanup()
```
Comment (dgozman) [inline playwright/_impl/_browser_type.py:232]: Bring this back?

File: playwright/_impl/_browser_type.py [hunk 14]
```diff
@@ -181,59 +179,50 @@ async def connect_over_cdp(
...
 
-        browser._set_browser_type(self)
+        pipe.on("closed", pipe_closed)
+
+        async def on_message(message: Dict) -> None:
+            try:
+                await pipe.send(message)
+            except Exception:
+                if not browser.is_connected:
+                    pipe_closed()
```
Comment (dgozman) [inline playwright/_impl/_browser_type.py:216]: Do this unconditionally? That's what JS does.

File: playwright/_impl/_browser_type.py [hunk 15]
```diff
@@ -181,59 +179,50 @@ async def connect_over_cdp(
...
+                raise
+
+        connection.on_message = on_message
+        pipe.on("message", connection.dispatch)
+        await connection.init()
+        playwright = await connection.playwright_future
+        browser: Browser = from_channel(playwright._initializer["preLaunchedBrowser"])
+        browser._is_remote = True
+        browser._should_close_connection_on_close = True
```
Comment (dgozman) [inline playwright/_impl/_browser_type.py:225]: JS has this line:
```js
browser.on(Events.Browser.Disconnected, closePipe);
```

File: playwright/_impl/_browser_type.py [hunk 16]
```diff
@@ -181,59 +179,50 @@ async def connect_over_cdp(
 
     async def connect(
         self,
-        ws_endpoint: str,
+        wsEndpoint: str,
         timeout: float = None,
```
Comment (dgozman) [inline playwright/_impl/_browser_type.py:183]: It seems like we ignore the timeout in the `await playwright_future` part.

File: playwright/_impl/_browser_type.py [hunk 17]
```diff
@@ -181,59 +179,50 @@ async def connect_over_cdp(
...
+            try:
+                await pipe.send(message)
+            except Exception:
+                if not browser.is_connected:
+                    pipe_closed()
+                raise
+
+        connection.on_message = on_message
+        pipe.on("message", connection.dispatch)
+        await connection.init()
```
Comment (dgozman) [inline playwright/_impl/_browser_type.py:221]: This and next lines should race against the timeout.

File: playwright/async_api/_context_manager.py [hunk 18]
```diff
@@ -25,30 +25,29 @@
...
         playwright.stop = self.__aexit__  # type: ignore
         return playwright
 
     async def start(self) -> AsyncPlaywright:
         return await self.__aenter__()
 
     async def __aexit__(self, *args: Any) -> None:
-        await self._connection.stop_async()
+        self._transport.request_stop()
+        await self._transport.wait_until_stopped()
```
Comment (dgozman) [inline playwright/async_api/_context_manager.py:53]: This probably needs `connection.cleanup()`

File: playwright/sync_api/_context_manager.py [hunk 19]
```diff
@@ -94,7 +99,8 @@ def start(self) -> SyncPlaywright:
         return self.__enter__()
 
     def __exit__(self, *args: Any) -> None:
-        self._connection.stop_sync()
+        self._transport.request_stop()
+        self._dispatcher_fiber.switch()
```
Comment (dgozman) [inline playwright/sync_api/_context_manager.py:103]: This probably needs `connection.cleanup()`

File: playwright/_impl/_object_factory.py [hunk 20]
```diff
@@ -40,7 +41,7 @@ class DummyObject(ChannelOwner):
     def __init__(
         self, parent: ChannelOwner, type: str, guid: str, initializer: Dict
     ) -> None:
-        super().__init__(parent, type, guid, initializer)
+        pass
```
Comment (dgozman) [inline playwright/_impl/_object_factory.py:44]: Why this change?

File: playwright/_impl/_connection.py [hunk 21]
```diff
@@ -205,38 +206,17 @@ def local_utils(self) -> "LocalUtils":
...
 
-    async def run_as_sync(self) -> None:
+    async def init_sync(self) -> None:
         self._is_sync = True
-        await self.run()
+        await self.init()
 
-    async def run(self) -> None:
-        self._loop = asyncio.get_running_loop()
+    async def init(self) -> None:
```
Comment (dgozman) [inline playwright/_impl/_connection.py:213]: This seems to be a sync method. Perhaps it could also be merged into the constructor.

File: playwright/_impl/_connection.py [hunk 22]
```diff
@@ -205,38 +206,17 @@ def local_utils(self) -> "LocalUtils":
...
-        self._loop.run_until_complete(self._transport.wait_until_stopped())
-        self.cleanup()
-
-    async def stop_async(self) -> None:
-        self._transport.request_stop()
-        await self._transport.wait_until_stopped()
-        self.cleanup()
-
-    def cleanup(self) -> None:
-        if self._init_task and not self._init_task.done():
```
Comment (dgozman) [inline playwright/_impl/_connection.py:235]: This seems like some good code.

**Review by dgozman — COMMENTED** [review]
Let's split this PR into ten 😄

**Comment by rwoll** [comment]
Closing for now—we should try splitting this up to make it easier to review. Currently, it's really challenging to understand what's going on.
